### ISSUE TITLE
Improvements related to recovery on the standby side (PG12 or higher)

### ISSUE THREAD

**Issue by MoonInsung** [ossc-db/pg_rman#153]
Improvements related to recovery on the standby side (PG12 or higher)
=============

What are the problems?
-------------
From PostgreSQL12 or later, recovery or SR standby side does not use recovery.conf, but uses *.signal file.
If standby.signal exists on the standby side and attempts to recovery using pg_rman, recovery.signal is created, and two signal files exist in one cluster.
There is no big problem because standby.signal is used preferentially in the PostgreSQL(it is written in the PostgreSQL manual). However, since an unnecessary *.signal is created, a solution is needed.

Cause of the problem
-------------
In the case of PostgreSQL11 and under version, the SR and recovery are set using recovery.conf.
Since PostgreSQL12 or higher version uses a separate file(recovery.signal, standby.signal), this problem has occurred.

Solutions
-------------
There are two ideas for the solution to me.
```
1. Forced recovery operation fails when standby.signal exists
When executing the recovery configuration of pg_rman(run to command pg_rman recovery), check whether standby.signal exists in the database cluster, an error occurs if the standby.signal file exists.

2. Operation of the --force option
When standby.signal exists When an option such as --force is added
How to delete standby.signal and add recovery.signal with a simple warning message
```

Note
-------------
**Basically, pg_rman does not support recovery on the standby side.**

Is there a better way?

Best regards.
Moon.

**Comment by mikecaat** [comment]
Thanks for reporting! 

I think the behavior isn't expected because the document said the following.

> Basically, the backup from standby-site is used for restoring on MASTER. pg_rman doesn’t treat the backup as restoring on standby automatically.  

If to obey the above description, I think to delete standby.signal is natural behavior without "--force" option of 1's solution. So, I think it's more like a bug than adding a feature, though? I made a PoC patch. Please see the above branch.

If to add a feature, I think to support restoring for standby's database cluster is better. There are two use-cases to restore from the backup data, which got from standby-side. But, Now pg_rman supports A only. Of course, a user can modify the data directory for B after restoring.

A. To restore for standby's database cluster.
B. To restore for new master's database cluster.

Since we need to handle the above two situations, it's not enough to implement --force option. For example, we need a option which has multiple choices like "--mode { primary | standby | none}". But, I don't have a strong opinion to add the option. I wonder that the solution to remove standby.signal is enough.

**Comment by MoonInsung** [comment]
Dear mikecaat.
thank you for the reply.

Unfortunately, in the case of Solution 1, it did not mean that is deleted option the 'standby.signal' but that the recovery operation failed.
The reason is that pg_rman basically defines that recovery is not possible on the stnadby side.

As explained in pseudocode, it is as follows.
```
recovery()
{
  if( stat(../standby.siganl,...))
  {
    ERR();
  }
}
```
If you delete the standby.signal, a side effect may occur because it is not the behavior defined by pg_rman so far.

In the case of Solution 2, which additionally removes the standby.signal, I thought that it was a prerequisite to cancel the SR configuration and make it work alone.
In other words, I thought that recovering the standby has the same meaning as NOT SR configuration.
So I meant a forced action like --force.

However, this is also different from pg_rman, which has been running so far, so it seems that there are several things to consider.

Therefore, it is recommended to use solution 1 to cause recovery to fail if standby.signal exists.
How does that sound?

Best regards.
Moon.

**Comment by mikecaat** [comment]
Thanks for your comments!

> The reason is that pg_rman basically defines that recovery is not possible on the stnadby side.

I think "basically" is the point. A user can start on the standby side after restoring if he/she modifies the restored data directory 's files. So, to reject if the "standby.signal" exists seems to be overdoing solution for me. If doing so, a user can't restore from the data backuped from the standby-side, which means to restore from the standby-side is not meaningless.

> If you delete the standby.signal, a side effect may occur because it is not the behavior defined by pg_rman so far.

In my understanding is correct, I think this behavior defined. The reason is that  

* Before PG11, pg_rman removes "standby_mode" in recovery.conf. This means that even if user backup from standby-side, the postgres start as a master at next startup
* The document said the following. This means that to remove "standby.signal" is the pg_rman's behavior because it restores on MASTER, doesn't it?  

> Basically, the backup from standby-site is used for restoring on MASTER. pg_rman doesn’t treat the backup as restoring on standby automatically.

Regards,

**Comment by MoonInsung** [comment]
Dear mikecaat.
Thank you for responding to my opinion.

Sorry. Like the comments and manuals, you provided
1. Recovery on the standby side is not **_NEVER_** impossible.
2. Recovery is possible on the standby side, but the SR configuration is released.
It was like this. Sorry for my misunderstanding.

So, I think the comments and PoCs you gave are in the right direction.
And when you send a PR, it would add a caution by emphasizing, "Be careful as the SR configuration is lost if you try to recover from the standby side" in the manual.

Best regards.
Moon.

**Comment by mikecaat** [comment]
Dear @MoonInsung 
Thanks! Yes, my understanding is the same as yours.
OK, I'll add the document.

BTY, I worry that how to test using the standby side...
Please let me know if you have any good ideas.

Regards,

**Comment by MoonInsung** [comment]
Dear mikecaat.

> BTY, I worry that how to test using the standby side...
> Please let me know if you have any good ideas.

How about referring to the regression test below for the standby side test?
```
sql/backup_from_standby.sh
```
Maybe, I think it will probably help.

Best regards.
Moon.

**Comment by mikecaat** [comment]
Dear @MoonInsung 
Great! Thanks, I'll reference it.
